doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,409 @@
+Python code. It will usually be source code.
+
+For example, if you run this file through Codegen:
+
+::
+
+    /* This is my C file. */
+    ...
+    /**
+     * @code{.codegen}
+     * fnames = ['DoSomething', 'DoAnotherThing', 'DoLastThing']
+     * for fn in fnames:
+     *     codegen.outl("void %s();" % fn)
+     * @endcode{.codegen}
+     */
+    /** @code{.codeins}@endcode */
+    ...
+
+it will come out like this:
+
+::
+
+    /* This is my C file. */
+    ...
+    /**
+     * @code{.codegen}
+     * fnames = ['DoSomething', 'DoAnotherThing', 'DoLastThing']
+     * for fn in fnames:
+     *     codegen.outl("void %s();" % fn)
+     * @endcode{.codegen}
+     */
+    void DoSomething();
+    void DoAnotherThing();
+    void DoLastThing();
+    /** @code{.codeins}@endcode */
+    ...
+
+Lines with @code{.codegen} or @code{.codeins}@endcode are marker lines.
+The lines between @code{.codegen} and @endcode{.codegen} are the generator
+Python code. The lines between @endcode{.codegen} and
+@code{.codeins}@endcode are the output from the generator.
+
+When Codegen runs, it discards the last generated Python output, executes the
+generator Python code, and writes its generated output into the file. All text
+lines outside of the special markers are passed through unchanged.
+
+The Codegen marker lines can contain any text in addition to the marker tokens.
+This makes it possible to hide the generator Python code from the source file.
+
+In the sample above, the entire chunk of Python code is a C comment, so the
+Python code can be left in place while the file is treated as C code.
+
+The codegen module
+******************
+
+A module called codegen provides the functions you call to produce output into
+your file. The functions are:
+
+.. function:: codegen.out(sOut=’’ [, dedent=False][, trimblanklines=False])
+
+    Writes text to the output.
+
+    :param sOut: The string to write to the output.
+    :param dedent: If dedent is True, then common initial white space is
+                   removed from the lines in sOut before adding them to the
+                   output.
+    :param trimblanklines: If trimblanklines is True,
+                           then an initial and trailing blank line are removed
+                           from sOut before adding them to the output.
+
+    ``dedent`` and ``trimblanklines`` make it easier to use
+    multi-line strings, and they are only are useful for multi-line strings:
+
+    ::
+
+        codegen.out("""
+            These are lines I
+            want to write into my source file.
+        """, dedent=True, trimblanklines=True)
+
+.. function:: codegen.outl
+
+    Same as codegen.out, but adds a trailing newline.
+
+.. function:: codegen.msg(msg)
+
+    Prints msg to stdout with a “Message: ” prefix.
+
+.. function:: codegen.error(msg)
+
+    Raises an exception with msg as the text. No traceback is included, so
+    that non-Python programmers using your code generators won’t be scared.
```

Comment:
I don't think being scared is an issue so I'd just write this as:

    No Python exception traceback is written.